rfcs/20250428-graph-api-sdpa-training/README.md

```diff
@@ -0,0 +1,563 @@
+# Supporting SDPA Training in Graph API
+
+## Background
+
+Scaled Dot-Product Attention (SDPA), introduced in [1], is a fundamental
+operation in Transformer blocks and serves as the backbone for numerous language
+and generative models (e.g., BERT, Stable Diffusion, GPT, DeepSeek).
+
+oneDNN supports SDPA inference and its optimization through the Graph API by
+defining an SDPA graph, partitioning it, and optimizing the underlying kernels.
+The floating-point SDPA — supporting `f32`, `bf16`, or `f16` — is defined as
+shown below:
+
+![oneDNN Graph SDPA pattern](images/onednn_graph_sdpa.png)
+
+The blue nodes are required when defining an SDPA pattern, while the brown nodes
+are optional. The general pattern computes `MatMul-SoftMax-MatMul`, as
+illustrated by the blue nodes. Given input sequences $Q$, $K$, and $V$:
+
+> $$
+> Q \in \mathbb{R}^{B \times H_{q} \times S_{q} \times D_{qk}}
+> $$
+> $$
+> K \in \mathbb{R}^{B \times H_{k} \times S_{kv} \times D_{qk}}
+> $$
+> $$
+> V \in \mathbb{R}^{B \times H_{v} \times S_{kv} \times D_{v}}
+> $$
+
+where:
+
+- $B$: Batch size
+- $H_{q}$: Number of query heads
+- $H_{k}$: Number of key heads
+- $H_{v}$: Number of value heads
+- $S_{q}$: Sequence length of the query
+- $S_{kv}$: Sequence length of the key and value
+- $D_{qk}$: Embedding dimension per head of query and key
+- $D_{v}$: Embedding dimension per head of value
+
+The attention output $O$ is computed as:
+
+> $$
+> S = Q \cdot K^T \in \mathbb{R}^{B \times H_{q} \times S_{q} \times S_{kv}}
+> $$
+> $$
+> P = f_{softmax}(S) \in \mathbb{R}^{B \times H_{q} \times S_{q} \times S_{kv}}
+> $$
+> $$
+> O = P \cdot V \in \mathbb{R}^{B \times H_{q} \times S_{q} \times D_{v}}
+> $$
+
+The standard implementation materializes the matrices $S$ and $P$ in
+high-bandwidth memory (HBM), which requires $O(S^2)$ memory. This becomes
+challenging for long sequence scenarios. To address this, oneDNN supports the
+FlashAttention 2 algorithm [2], while the reference implementation still follows
+the standard approach.
+
+For additional opt-in features, such as attention scaling and masking, the
+computations are adjusted accordingly:
+
+> $$
+> S_{scaled} = f_{scale}(S) \in \mathbb{R}^{B \times H_{q} \times S_{q} \times S_{kv}}
+> $$
+> $$
+> S_{masked} = f_{mask}(S_{scaled}) \in \mathbb{R}^{B \times H_{q} \times S_{q} \times S_{kv}}
+> $$
+
+The following sections explore how to support SDPA training in the oneDNN Graph
+API. We will start from `f32` data type.
```

Comment:
Just to clarify, do you plan to cover bf16, f16, or even fp8 training in this document?

Reply:
Plan to start from f32, I add the clarification at the section of implementation effort. Thanks.